R/setup.R

```diff
@@ -1,44 +1,117 @@
+                              user = github_user()$login,
+                              repo = github_info(path = path)$name,
+                              endpoint = get_endpoint(),
+                              quiet = FALSE) {
 
-  # authenticate on github and travis and set up keys/vars
+  auth_github()
+  auth_travis()
 
   # generate deploy key pair
-  key <- openssl::rsa_keygen()  # TOOD: num bits?
+  key <- openssl::rsa_keygen() # TOOD: num bits?
 
   # encrypt private key using tempkey and iv
   pub_key <- get_public_key(key)
   private_key <- encode_private_key(key)
 
-  # add to GitHub first, because this can fail because of missing org permissions
-  title <- paste0("travis+tic", if (repo == travis_repo) "" else (paste0(" for ", repo)))
-  github_add_key(pub_key, title = title, info = info)
+  # Github deploy key ----------------------------------------------------------
 
-  travis_set_var("id_rsa", private_key, public = FALSE, repo = travis_repo)
+  # query deploy key
+  if (!quiet) {
+    cli::cli_alert_info("Querying Github deploy keys from repo.")
+  }
+  gh_keys <- gh::gh("/repos/:owner/:repo/keys",
+    owner = github_info(path = path)$owner$login, repo = repo
+  )
+
+  if (!gh_keys[1] == "") {
+    gh_keys_names <- gh_keys %>%
+      purrr::map_chr(~ .x$title)
+
+    # delete old keys with no endpoint spec
+    # this helps to avoid having unused keys stored
+    old_keys <- gh_keys_names %>%
+      purrr::map_lgl(~ .x == "Deploy key for Travis CI" | .x == "travis+tic")
+    if (any(old_keys == TRUE)) {
+      purrr::walk(gh_keys[old_keys], ~
+      gh::gh("DELETE /repos/:owner/:repo/keys/:key_id",
+        owner = github_info(path = path)$owner$login,
+        repo = repo,
+        key_id = .x$id
+      ))
+      if (!quiet) {
+        cli::cli_alert_info("Deleted unused old Travis deploy key(s) from
+                          Github repo.", wrap = TRUE)
+      }
+    }
+
+    # check if key(s) exists
+    if (any(gh_keys_names %in% sprintf(
+      "Deploy key for Travis CI (%s)",
+      endpoint
+    ))) {
+      return(cli::cli_alert("Deploy key for Travis CI ({.code {endpoint}})
+      already present in {.file ~/.travis/config.yml}. No action required.",
+        wrap = TRUE
+      ))
+    }
+  }
+
+  # add to GitHub first, because this can fail because of missing org
+  # permissions
+  title <- sprintf("Deploy key for Travis CI (%s)", endpoint)
+  github_add_key(pubkey = pub_key, user = user, repo = repo, title = title)
 
-  cli::cat_bullet(
-    bullet = "tick", bullet_col = "green",
-    paste0(
-      "Successfully added private deploy key to ",
-      travis_repo, " as secure environment variable id_rsa to Travis CI."
+  # env var 'id_rsa' on Travis -------------------------------------------------
+
+  # check if id_rsa already exists on Travis
+  idrsa <- travis_get_vars(
+    repo = github_repo(path = path),
+    endpoint = endpoint
+  ) %>%
+    purrr::map_lgl(~ .x$name == "id_rsa") %>%
```

Comment:
The `"id_rsa"` name comes from tic. This constant here is a hidden dependency. How to resolve?

Reply:
Hm, because of `tic::ci_can_push()`.

I am wondering if we can also deploy on Travis without the SSH key pair? Haven't yet tried hard but since I also do not need it on Circle.
I remember it being troublesome because Travis relies on `dpl` for deployment.

Nevertheless, this should probably go into its own issue.

Reply:
https://github.com/ropenscilabs/travis/pull/115#discussion_r357943501